[MXNET-50] Scala Inference APIs

nswamy opened the pull request:
## Description ##
https://issues.apache.org/jira/browse/MXNET-50
MXNet Scala Inference Library. 

## Checklist ##
### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

marcoabreu commented:
Would it make sense to have this package under the org.apache namespace since it seems like an entirely new aspect? This might allow us to make a quicker transition, away from dmlc.

nswamy commented:
@marcoabreu, this package needs to pull in the core package under dmlc, so i prefer to keep it under dmlc. I would prefer to have a separate discussion to see how we want to manage versioning of language bindings for MXNet. Currently, It does not make sense to me that a API change in one of the language bindings trigger whole new version for entire MXNet. Also Scala packages have not been published to maven for a while now and the versions are maintained in Maven. we have to address this as well.

nswamy commented:
Unit Tests are failing for a requirement(number of dimensions in shape match the number of elements in layout) I put into the DataDesc. I will fix the unit tests and update the PR.

nswamy commented:
@lupesko please review.

yzhliu commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/MXNetHandler.scala:
```diff
@@ -0,0 +1,82 @@
+    val task = new Callable[T] {
+      override def call(): T = {
+        // scalastyle:off println
+        println("threadId: %s".format(Thread.currentThread().getId()))
```
better to use Log

nswamy replied:
this for testing, removed the print

yzhliu commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/MXNetHandler.scala:
```diff
@@ -0,0 +1,82 @@
+
+}
+
+object MXNetSingleThreadHandler extends MXNetOneThreadPerModelHandler {
```
why do we need an `object` extends something? any `static` members to be accessed?

nswamy replied:
I want to create a singleton object with only thread which will be used by default.

yzhliu commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/MXNetHandler.scala:
```diff
@@ -0,0 +1,82 @@
+    }
+  }
+
+  override val executor: ExecutorService = Executors.newFixedThreadPool(10, threadFactory)
```
make `10` configurable through constructor.

nswamy replied:
I want this to be configurable, the value 10 was for testing, made this configurable now.

yzhliu commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/MXNetHandler.scala:
```diff
@@ -0,0 +1,82 @@
+      result.get()
+    }
+    catch {
+      case e: ExecutionException => throw e.getCause()
```
add `throw ExecutionException` to the signature of this method.
place `catch` the same line of `}`

yzhliu commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/MXNetHandler.scala:
```diff
@@ -0,0 +1,82 @@
+    if (handlerType == MXNetHandlerType.OneThreadPerModelHandler) {
+      new MXNetOneThreadPerModelHandler
+    }
+    else {
```
place `else` the same line of `}`

yzhliu commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/MXNetHandler.scala:
```diff
@@ -0,0 +1,82 @@
+
+  def execute[T](f: => T): T
+
+  val executor: ExecutorService
```
`private`

calumleslie replied:
Arguably this should not be on the trait at all.

nswamy replied:
I want to be able to expose the executor if at all I find that all NDArray creations have to go throw the same thread.

yzhliu commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/package.scala:
```diff
@@ -0,0 +1,24 @@
+import ml.dmlc.mxnet.infer.MXNetHandlerType.MXNetHandlerType
+
+package object infer {
+  private[mxnet] val handlerType: MXNetHandlerType = MXNetHandlerType.SingleThreadHandler
```
can we change this variable anywhere?

CodingCat replied:
should we make it user-configurable?

nswamy replied:
I will expose this as a property if I find that MXNet can be called from multiple threads.
, like said in other places I am going with the assumption that MXNet needs to be called from the same thread through out the lifetime of the process otherwise it seg-faults.

CodingCat commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/PredictBase.scala:
```diff
@@ -0,0 +1,200 @@
+    return mod
+
+  }
+
```
extra empty line

CodingCat commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/PredictBase.scala:
```diff
@@ -0,0 +1,200 @@
+    val mod = mxNetHandler.execute(Module.loadCheckpoint(modelPathPrefix, 0))
+    mxNetHandler.execute(mod.bind(inputDescriptors, forTraining = false))
+    return mod
+
```
extra empty line

CodingCat commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/PredictBase.scala:
```diff
@@ -0,0 +1,200 @@
+      mxNetHandler.execute(mod.bind(iDescriptors, outputDescriptors, forceRebind = true,
+        forTraining = false))
+    }
+
```
extra empty line

CodingCat commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/PredictBase.scala:
```diff
@@ -0,0 +1,200 @@
+
+    // rebind with the new batchSize
+    if (batchSize != inputBatchSize) {
+
```
extra empty line

CodingCat commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/PredictBase.scala:
```diff
@@ -0,0 +1,200 @@
+
+    // Shape validation, remove this when backend throws better error messages.
+    for((i, d) <- inputBatch.zip(iDescriptors)) {
+
```
extra empty line

CodingCat commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/PredictBase.scala:
```diff
@@ -0,0 +1,200 @@
+    }
+
+    result
+
```
extra empty line

CodingCat commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/PredictBase.scala:
```diff
@@ -0,0 +1,200 @@
+    if (batchSize != 1) {
+      mxNetHandler.execute(mod.bind(inputDescriptors, forTraining = false, forceRebind = true))
+    }
+
```
extra empty line

nswamy replied:
Is it standard to not have blank lines, I tend to put blank lines so that it is more readable.

CodingCat commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/PredictBase.scala:
```diff
@@ -0,0 +1,200 @@
+
+    for((i, d) <- input.zip(inputDescriptors)) {
+      val shape = d.shape.toVector.patch(from = batchIndex, patch = Vector(1), replaced = 1)
+
```
extra empty line

CodingCat replied:
please check all the other places

CodingCat commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/PredictBase.scala:
```diff
@@ -0,0 +1,200 @@
+  /**
+    * This method will take input as IndexedSeq one dimensional arrays and creates
+    * NDArray needed for inference. The array will be reshaped based on the input descriptors.
+    * @param input: A IndexedSequence of Java one-dimensional array, An IndexedSequence is
```
Java?

CodingCat commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/PredictBase.scala:
```diff
@@ -0,0 +1,200 @@
+    for((i, d) <- input.zip(inputDescriptors)) {
+      val shape = d.shape.toVector.patch(from = batchIndex, patch = Vector(1), replaced = 1)
+
+      inputND += mxNetHandler.execute(NDArray.array(i, Shape(shape)))
```
I am a bit confused here, `NDArray.array(i, Shape(shape))` is executed by a new thread? 

and it's actually blocking here until the function is finished? why we need to do this?

calumleslie replied:
Calls to MXNet's native library are not safe to call from more than one thread per program execution (unless that has been changed). This is just how MXNet works at present. We [described this on the discussion site at one point](https://discuss.mxnet.io/t/fixing-thread-safety-issues-in-scala-library/236).

It's actually not entirely known if some calls are safe and some are not but we've found the best stability from single-threading all calls to MXNet.

nswamy replied:
I am going by Calum's findings here

calumleslie commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/PredictBase.scala:
```diff
@@ -0,0 +1,200 @@
+
+    for((i, d) <- input.zip(inputDescriptors)) {
+      require (i.length == d.shape.product/batchSize, "number of elements:" +
+        " %d in the input does not match the shape:%s".format( i.length, d.shape.toString()))
```
Minor: The use of spaces within parentheses here is inconsistent.

CodingCat commented:
Hi, the community has passed to vote about associating the code changes with JIRA (https://lists.apache.org/thread.html/ab22cf0e35f1bce2c3bf3bec2bc5b85a9583a3fe7fd56ba1bbade55f@%3Cdev.mxnet.apache.org%3E)

We have updated the guidelines for contributors in https://cwiki.apache.org/confluence/display/MXNET/Development+Process, please ensure that you have created a JIRA at https://issues.apache.org/jira/projects/MXNET/issues/ to describe your work in this pull request and include the JIRA title in your PR as [MXNET-xxxx] your title where MXNET-xxxx is the JIRA id

Thanks!

nswamy commented:
@CodingCat @yzhliu @calumleslie 
I have addressed most of your comments and added more code. Unfortunately i force pushed to my branch github is now not showing old comments, If I have missed any let me know.
I have also added unit tests.

@CodingCat 
I don't know the specific benefits of using Scala Concurrency features, I think this suffices for now and I'll change it to Scala Concurrency as I learn and improve this in the next iteration.

@calumleslie 
For your comment on using checking if it is already on a thread that can run MXNet, I didn't have any issues even when on the same thread. May be I did not understand the problem, can you please elaborate.

yzhliu commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/Classifier.scala:
```diff
@@ -0,0 +1,167 @@
+
+  val synsetFilePath = getSynsetFilePath(modelPathPrefix)
+
+  val synset = readSynsetFile(synsetFilePath)
```
I think it is better to make synset optional

nswamy replied:
Classifier is just taking the predictor and using Synset to map the labels, this is the only difference between the predictor/classifier

yzhliu commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/Predictor.scala:
```diff
@@ -0,0 +1,188 @@
+    resultND
+  }
+
+  def loadModule(): Module = {
```
I feel it is better to rely on `Symbol` and `Executor` directly. `Executor` is much cleaner and cannot run into any memory leak before we fix the GC problem.

nswamy replied:
wouldn't this just be bringing parts of the `Module` code into this class ?

nswamy replied:
discussed offline with @yzhliu, Module is necessary to support GPUs.

yzhliu replied:
this function should be `private`, right?

nswamy commented on scala-package/infer/pom.xml:
```diff
@@ -0,0 +1,124 @@
+                </plugins>
+            </build>
+        </profile>
+    </profiles>
```
I wasn't able inherit and make the tests run. the tests needs the jar to be built

nswamy commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/MXNetHandler.scala:
```diff
@@ -0,0 +1,82 @@
+
+class MXNetOneThreadPerModelHandler extends MXNetHandler {
+
+  private val threadFactory = new ThreadFactory {
```
I'll learn about it and let me do that in the next iteration

nswamy commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/PredictBase.scala:
```diff
@@ -0,0 +1,200 @@
+
+/**
+  * Base Trait for MXNNet Predictor classes.
+  */
```
I wonder why scala-style is not already doing that.

nswamy commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/MXNetHandler.scala:
```diff
@@ -0,0 +1,82 @@
+
+  type MXNetHandlerType = Value
+  val SingleThreadHandler = Value("MXNetSingleThreadHandler")
+  val OneThreadPerModelHandler = Value("MXNetOneThreadPerModelHandler")
```
To be able to support in the future or if I find that its OK to run one-thread-per-model

nswamy commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/MXNetHandler.scala:
```diff
@@ -0,0 +1,82 @@
+
+  override val executor: ExecutorService = Executors.newFixedThreadPool(10, threadFactory)
+
+  override def execute[T](f: => T): T = {
```
I did not encounter this when i unit tested by running on the same thread. However I fixed this based on your implementation.

nswamy commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/Classifier.scala:
```diff
@@ -0,0 +1,167 @@
+    * @return IndexedSequence of (Label, Score) tuples.
+    */
+  def classify(input: IndexedSeq[Array[Float]],
+               topK: Option[Int] = None): List[(String, Float)]
```
Ok, changed the return type to IndexedSeq instead.

nswamy commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/Classifier.scala:
```diff
@@ -0,0 +1,170 @@
+    * @param input: Indexed Sequence of NDArrays
+    * @param topK: (Optional) How many top_k(sorting will be based on the last axis)
+    *             elements to return, if not passed returns unsorted output.
+    * @return Traversable Sequence of (Label, Score) tuple, Score will be in the form of NDArray
```
this remained from the previous version of the code. thanks for pointing. fixed

yzhliu replied:
sorry, have you fixed? I don't see from my end.

nswamy commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/Predictor.scala:
```diff
@@ -0,0 +1,197 @@
+   * This method will take input as IndexedSeq one dimensional arrays and creates
+   * NDArray needed for inference. The array will be reshaped based on the input descriptors.
+   * @param input: A IndexedSequence of Scala one-dimensional array, An IndexedSequence is
+   *             is needed when the model has more than one input/output
```
this for models that need more than 1 input, a crude example would be a model that takes 2 different images

nswamy commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/Predictor.scala:
```diff
@@ -0,0 +1,197 @@
+
+  protected[mxnet] val mxNetHandler = MXNetHandler()
+
+  protected[mxnet] val mod = loadModule()
```
if i don't make it package[infer] private. I won't be able to mock and unit test easily.

nswamy commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/MXNetHandler.scala:
```diff
@@ -0,0 +1,94 @@
+  override def execute[T](f: => T): T = {
+
+    if (Thread.currentThread() eq creatorThread) {
+      f
```
this doesn't work since it does not take any parameters, so I added a return to be more explicit

calumleslie commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/MXNetHandler.scala:
```diff
@@ -0,0 +1,95 @@
+  override val executor: ExecutorService =
+    Executors.newFixedThreadPool(numThreads.get, threadFactory)
+
+  private val creatorThread = executor.submit(new Callable[Thread] {
```
I'm not sure if `creatorThread` is going to have a meaningful value when the size of the threadpool is not 1.

Given that the size of the thread pool _must always_ be 1 at the moment, maybe it'd be easier to just avoid having to handle multiple threads?

calumleslie commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/MXNetHandler.scala:
```diff
@@ -0,0 +1,95 @@
+      try {
+        result.get()
+      } catch {
+        case e: Exception => throw e.getCause()
```
You should probably only be doing this for ExecutionException. This method can also throw InterruptedException which is not guaranteed to have a cause.

I also think that you should comment explaining why you're unwrapping this exception here.

nswamy replied:
you might also get other exceptions, I think it makes sense to catch all Exceptions. I will throw InterruptedException as is separately.

calumleslie replied:
I believe this API should only ever throw ExecutionException or InterruptedException but the point was the only type of exception you can safely call "getCause" on is the former. You probably just want all other exceptions to be propagated as is which you can do by just not catching them.

calumleslie commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/MXNetHandler.scala:
```diff
@@ -0,0 +1,95 @@
+  val OneThreadPerModelHandler = Value("MXNetOneThreadPerModelHandler")
+}
+
+private[infer] class MXNetThreadPoolHandler(numThreads: Option[Int] = Some(1))
```
`numThreads` is an `Option[Int]` but you always unconditionally dereference it. Is `None` a valid value? Could you just accept an `Int` here?

nswamy replied:
good point, regardless of whether an Option or just a Int, I needed a validation of int > 0. I changed it to a Int and added validation.

calumleslie commented on scala-package/infer/src/main/scala/ml/dmlc/mxnet/infer/MXNetHandler.scala:
```diff
@@ -0,0 +1,95 @@
+
+  private val threadFactory = new ThreadFactory {
+
+    override def newThread(r: Runnable): Thread = new Thread(r) {
```
This will create all threads with the same name, which will make them difficult (or in some contexts impossible) to tell apart. If you're going to continue to support multiple threads consider incrementing a counter whenever you vend a thread and appending the number to the name (this is the behaviour of Guava's [ThreadFactoryBuilder](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/util/concurrent/ThreadFactoryBuilder.html) for example).

calumleslie commented:
@nswamy Sorry it took me a while to get back I've been off ill. Also I don't really use GitHub so often miss the notifications! I've added some more comments on some of the implementation here.

>  For your comment on using checking if it is already on a thread that can run MXNet, I didn't have any issues even when on the same thread. May be I did not understand the problem, can you please elaborate.

It looks like you fixed this with the `creatorThread` stuff :)
